proposals/0021-root-signature-in-clang.md

```diff
@@ -0,0 +1,205 @@
+<!-- {% raw %} -->
+
+# Implementation of RootSignature in clang
+
+* Proposal: [0021](0021-root-signature-in-clang.md)
+* Author(s): [Xiang Li](https://github.com/python3kgae)
+* Sponsor: TBD
+* Status: **Under Consideration**
+
+
+## Introduction
+
+In HLSL, the [root signature](https://learn.microsoft.com/en-us/windows/win32/direct3d12/root-signatures)
+defines what types of resources are bound to the graphics pipeline.
+
+A root signature can be specified in HLSL as a [string](https://learn.microsoft.com/en-us/windows/win32/direct3d12/specifying-root-signatures-in-hlsl#an-example-hlsl-root-signature).
+The string contains a collection of comma-separated clauses that describe root
+signature constituent components.
+
+## Motivation
+
+To support root signature in clang-dxc.
+
+## Proposed solution
+
+The root signature in a compiler begins with a string in the root signature
+attribute and ends with a serialized root signature in the DXContainer.
+
+To report errors as early as possible, root signature string parsing should
+occur in Sema.
+
+To ensure that a used resource has a binding in the root signature, this
+information should be accessible in the backend to make sure legal dxil is
+generated.
+
+A new AST node HLSLRootSignatureDecl will be added to represent the root
+signature in the AST.
+Assuming the cost to parse root signature string is cheap,
+HLSLRootSignatureDecl will follow common Clang approach which only save the
+string in the AST.
+
+A root signature will be parsed twice, once in Sema for diagnostic and once in
+clang code generation for generating the serialized root signature.
+The first parsing will check register overlap and run on all root signatures
+in the translation unit.
+The second parsing will calculate the offset for serialized root signature and
+only run on the root signature for the entry function or local/global root
+signature which used in SubobjectToExportsAssociation.
+
+HLSLRootSignatureDecl will have method to parse the string for diagnostic and
+generate the SerializedRS mentioned above.
+
+A HLSLRootSignatureAttr will be created when meet RootSignature attribute in
+HLSL.
+
+For case compile to a standalone root signature blob, the
+HLSLRootSignatureAttr will be bind to a fake empty entry.
+
+In clang code generation, the HLSLRootSignatureAttr in AST will be translated
+into a global variable with struct type to express the layout and metadata to
+save things like static sampler, root flags, space and NumDescriptors in LLVM IR.
+
+CGHLSLRuntime will generate metadata to link the global variable as root
+signature for given entry function.
```

Comment:
What is this global variable? Are you expecting this is fully-serialized?

That would be unfortunate because it would mean any other programming language that wants to use the DirectX backend would also need to know how to serialize the binary representation of the root signature and embed it in the IR. It would be easier for other languages if they just needed to emit metadata to describe the root signature.

Reply:
Unless what saved in the metadata is the original root signature string, generate metadata is just add another representation which other languages need to generate like serialized root signature layout.

If it is OK, I think save the string in LLVM IR is easier.
But not sure about share parsing code between clang and llvm.

Reply:
No other language is going to use the root signature string that HLSL uses. That's a very HLSL-specific design decision.

Other languages might decide to do a structure-based interface, or (like Metal) a parameter-based interface. Forcing those frontends to generate the DirectX binary serialization or the HLSL string are both really awkward design decisions.

Reply:
We can expose struct like
  struct ParsedRootSignature {
    dxbc::RootSignature::ContainerRootSignatureDesc RSDesc;
    SmallVector<dxbc::RootSignature::ContainerRootParameter, 8> RSParameters;
    SmallVector<dxbc::RootSignature::StaticSamplerDesc, 8> StaticSamplers;
    SmallVector<dxbc::RootSignature::v_1_1::ContainerDescriptorRange, 8>
        DescriptorRanges;
    SmallVector<
        std::variant<dxbc::RootSignature::ContainerRootConstants,
                     dxbc::RootSignature::v_1_1::ContainerRootDescriptor,
                     dxbc::RootSignature::ContainerRootDescriptorTable>,
        8>
        RSParamExtras;
  };
And offer function to generate the representation for root signature in llvm IR from a ParsedRootSignature .

It will be easier than create metadata directly.

Reply:
No.
Updated the global variable to metadata.

Reply:
@llvm-beanz - ok to resolve this conversation?